### PR TITLE
[Docs] guide for query performance

### PR DESCRIPTION
**Pull request by MildTomato**

### CODE REVIEW

File: apps/docs/pages/guides/platform/performance.mdx [hunk 1]
```diff
@@ -8,6 +8,152 @@ export const meta = {
...
+
+This query will show you statistics about queries ordered by the cumulative total execution time. It shows the total time the query has spent running as well as the proportion of total execution time the query has taken up.
+
+Queries which are the most time consuming are not necessarily bad, you may have a very effiecient and frequently ran queries that end up taking a large total % time, but it can be useful to help spot queries that are taking up more time than they should.
+
+### Hit rate
+
+Generally for most applications a small percentage of data is accessed more regularly than the rest. To make sure that your regularly accessed data is available, Postgres tracks your data access patterns and keeps this in its [shared_buffers](https://www.postgresql.org/docs/15/runtime-config-resource.html#RUNTIME-CONFIG-RESOURCE-MEMORY) cache.
+
+Applications with lower cache hit rates generally perform more poorly since they have to hit the disk to get results rather than serving them from memory. Very poor hit rates can also cause you to burst past your [Disk I/O limits](https://supabase.com/docs/guides/platform/compute-add-ons#disk-io-bandwidth) causing significat performance issues.
```
[the remark this answers is not in the document]
Comment (KevinBrolly) [reply apps/docs/pages/guides/platform/performance.mdx:117]: Good bot

**Comment by MildTomato** [comment]
hehe https://github.com/supabase/supabase/pull/12555/commits/226679f2f004dd67af3ec02c57c81782fb413793

**Comment by MildTomato** [comment]
undrafting :D

File: apps/docs/pages/guides/platform/performance.mdx [hunk 2]
```diff
@@ -8,6 +8,152 @@ export const meta = {
 
 The Supabase platform automatically optimizes your Postgres database to take advantage of the compute resources of the tier your project is on. However, these optimizations are based on assumptions about the type of workflow the project is being utilized for, and it is likely that better results can be obtained by tuning the database for your particular workflow.
 
+## Examining Query Performance
+
+A major cause of poor database performance is unoptimized queries. The techniques on this page can help you identify and understand queries that take the most time and resources from your database.
```
Comment (kiwicopple) [inline apps/docs/pages/guides/platform/performance.mdx:13]:
```suggestion
Unoptimized queries are a major cause of poor database performance. The techniques on this page can help you identify and understand queries that take the most time and resources from your database.
```

(use active voice)

image

File: apps/docs/pages/guides/platform/performance.mdx [hunk 3]
```diff
@@ -8,6 +8,152 @@ export const meta = {
...
+
+Database performance is a large topic and many factors can contribute, but some of the most common causes of poor performance include:
```
Comment (kiwicopple) [inline apps/docs/pages/guides/platform/performance.mdx:15]:
```suggestion
Database performance is a large topic and many factors can contribute. Some of the most common causes of poor performance include:
```

File: apps/docs/pages/guides/platform/performance.mdx [hunk 4]
```diff
@@ -8,6 +8,152 @@ export const meta = {
...
+* Unused indexes causing slow `INSERT`, `UPDATE` and `DELETE` operations
+* Not enough compute resources, such as memory, causing your database to go to disk for results too often
+* Lock contention from multiple queries operating on highly utilized tables
+* Large amount of bloat on your tables causing poor query planning
+
+Thankfully there are solutions to all these issues, which we will cover in the following sections.
+
+### Postgres Cumulative Statistics system
+
+Postgres itself collects lots of data about its own operations using the [cumulative statistics system](https://www.postgresql.org/docs/current/monitoring-stats.html). In addition to this, every Supabase project has [the pg_stat_statements extension](https://www.postgresql.org/docs/14/pgstatstatements.html) enabled by default. This extension records query execution performance details and is the best way to find queries that take the most time to execute. This information can be combined with the Postgres query plan analyzer to develop more efficient queries.
```
Comment (kiwicopple) [inline apps/docs/pages/guides/platform/performance.mdx:29]:
```suggestion
Postgres collects data about its own operations using the [cumulative statistics system](https://www.postgresql.org/docs/current/monitoring-stats.html). In addition to this, every Supabase project has the [pg_stat_statements extension](/docs/guides/database/extensions/pg_stat_statements) enabled by default. This extension records query execution performance details and is the best way to find inefficient queries. This information can be combined with the Postgres query plan analyzer to develop more efficient queries.
```

File: apps/docs/pages/guides/platform/performance.mdx [hunk 5]
```diff
@@ -8,6 +8,152 @@ export const meta = {
...
+
+It is easy to use `pg_stat_statements` and other Postgres statistics tables to get a wealth of information about your database. Here are some example queries to get you started.
```
Comment (kiwicopple) [inline apps/docs/pages/guides/platform/performance.mdx:31]:
```suggestion
Here are some example queries to get you started.
```

Show don't tell

File: apps/docs/pages/guides/platform/performance.mdx [hunk 6]
```diff
@@ -8,6 +8,152 @@ export const meta = {
...
+
+from pg_stat_statements as statements
+  inner join pg_authid as auth on statements.userid = auth.oid
+order by
+  statements.calls desc
+limit
+  100;
+```
+
+This query will show you statistics about queries ordered by the number of times they have been called. It shows the role that ran the query, the number of times it has been called, the average number of rows returned, the cumulative total time the query has spent running as well as the min, max and mean query times.
```
Comment (kiwicopple) [inline apps/docs/pages/guides/platform/performance.mdx:60]:
```suggestion
This query shows:

- query statistics, ordered by the number of times each query has been executed
- the role that ran the query
- the number of times it has been called
- the average number of rows returned
- the cumulative total time the query has spent running 
- the min, max and mean query times.
```

File: apps/docs/pages/guides/platform/performance.mdx [hunk 7]
```diff
@@ -8,6 +8,152 @@ export const meta = {
...
+
+This can provide useful information about queries that you are running often. You should try and ensure that any queries that are being called often are performant and have good mean exeuction times. Queries which are showing as having high max or mean times and are being called often can be good candidates for optimisation.
```
Comment (kiwicopple) [inline apps/docs/pages/guides/platform/performance.mdx:62]:
```suggestion
This provides useful information about the queries you run most frequently. Queries that have high `max_time` or `mean_time` times and are being called often can be good candidates for optimization.
```

File: apps/docs/pages/guides/platform/performance.mdx (hunk 2 above)
Comment (MildTomato) [reply apps/docs/pages/guides/platform/performance.mdx:13]: 🚀

File: apps/docs/pages/guides/platform/performance.mdx (hunk 6 above)
Comment (MildTomato) [reply apps/docs/pages/guides/platform/performance.mdx:60]: nice!

File: studio/components/interfaces/SQLEditor/SQLEditor.constants.ts [hunk 8]
```diff
@@ -969,4 +969,106 @@ GRANT ALL ON TABLE next_auth.verification_tokens TO postgres;
 GRANT ALL ON TABLE next_auth.verification_tokens TO service_role;
 `.trim(),
   },
+  {
```
Comment (MildTomato) [inline studio/components/interfaces/SQLEditor/SQLEditor.constants.ts:972]: @joshenlim @alaister want to just check these are ok?

File: studio/components/interfaces/SQLEditor/SQLEditor.constants.ts [hunk 9]
```diff
@@ -969,4 +969,106 @@ GRANT ALL ON TABLE next_auth.verification_tokens TO postgres;
...
+    sql: `-- Most time consuming queries
+
+-- A limit of 100 has been added below
+
+select
+    auth.rolname,
+    statements.query,
+    statements.calls,
+    statements.total_exec_time + statements.total_plan_time as total_time,
+    to_char(((statements.total_exec_time + statements.total_plan_time)/sum(statements.total_exec_time + statements.total_plan_time) OVER()) * 100, 'FM90D0') || '%'  AS prop_total_time
```
Comment (alaister) [inline studio/components/interfaces/SQLEditor/SQLEditor.constants.ts:1018]: nit: uppercase `OVER` and `AS`

**Review by alaister — COMMENTED** [review]
Queries look good to me @MildTomato. Only one small nit:

File: studio/components/interfaces/SQLEditor/SQLEditor.constants.ts (hunk 9 above)
Comment (MildTomato) [reply studio/components/interfaces/SQLEditor/SQLEditor.constants.ts:1018]: should be lowercase you think?

Comment (MildTomato) [reply studio/components/interfaces/SQLEditor/SQLEditor.constants.ts:1018]: oh yea

Comment (alaister) [reply studio/components/interfaces/SQLEditor/SQLEditor.constants.ts:1018]: yep

Comment (MildTomato) [reply studio/components/interfaces/SQLEditor/SQLEditor.constants.ts:1018]:
```suggestion
statements.total_plan_time) over()) * 100, 'FM90D0') || '%' as prop_total_time
```

Comment (MildTomato) [reply studio/components/interfaces/SQLEditor/SQLEditor.constants.ts:1018]:
```suggestion
    to_char(((statements.total_exec_time + statements.total_plan_time)/sum(statements.total_exec_time + statements.total_plan_time) over()) * 100, 'FM90D0') || '%' as prop_total_time
```

File: apps/docs/pages/guides/platform/performance.mdx [hunk 10]
```diff
@@ -8,6 +8,159 @@ export const meta = {
...
+
+Thankfully there are solutions to all these issues, which we will cover in the following sections.
+
+### Postgres Cumulative Statistics system
+
+Postgres collects data about its own operations using the [cumulative statistics system](https://www.postgresql.org/docs/current/monitoring-stats.html). In addition to this, every Supabase project has the [pg_stat_statements extension](/docs/guides/database/extensions/pg_stat_statements) enabled by default. This extension records query execution performance details and is the best way to find inefficient queries. This information can be combined with the Postgres query plan analyzer to develop more efficient queries.
+
+Here are some example queries to get you started.
+
+Most frequently called queries:
```
Comment (MildTomato) [inline apps/docs/pages/guides/platform/performance.mdx:33]:
```suggestion
#### Most frequently called queries:
```

so we can link to it

File: apps/docs/pages/guides/platform/performance.mdx [hunk 11]
```diff
@@ -8,6 +8,159 @@ export const meta = {
...
+- query statistics, ordered by the number of times each query has been executed
+- the role that ran the query
+- the number of times it has been called
+- the average number of rows returned
+- the cumulative total time the query has spent running 
+- the min, max and mean query times.
+
+This provides useful information about the queries you run most frequently. Queries that have high `max_time` or `mean_time` times and are being called often can be good candidates for optimization.
+
+Slowest queries by execution time:
```
Comment (MildTomato) [inline apps/docs/pages/guides/platform/performance.mdx:71]:
```suggestion
#### Slowest queries by execution time:
```

so we can link to it

File: apps/docs/pages/guides/platform/performance.mdx [hunk 12]
```diff
@@ -8,6 +8,159 @@ export const meta = {
...
+    inner join pg_authid as auth on statements.userid = auth.oid
+  order by
+    max_time desc
+  limit
+    100;
+```
+
+This query will show you statistics about queries ordered by the maximum execution time. It is similar to the query above ordered by calls, but this one highlights outliers that may have high executions times. Queries which have high or mean execution times are good candidates for optimisation.
+
+Most time consuming queries:
```
Comment (MildTomato) [inline apps/docs/pages/guides/platform/performance.mdx:99]:
```suggestion
#### Most time consuming queries:
```

so we can link to it

File: apps/docs/pages/guides/platform/performance.mdx [hunk 13]
```diff
@@ -8,6 +8,159 @@ export const meta = {
...
+
+#### Most time consuming queries:
+
+```sql
+select
+  auth.rolname,
+  statements.query,
+  statements.calls,
+  statements.total_exec_time + statements.total_plan_time as total_time,
+  to_char(((statements.total_exec_time + statements.total_plan_time)/sum(statements.total_exec_time + statements.total_plan_time) OVER()) * 100, 'FM90D0') || '%'  AS prop_total_time
```
Comment (alaister) [inline apps/docs/pages/guides/platform/performance.mdx:107]:
```suggestion
  to_char(((statements.total_exec_time + statements.total_plan_time)/sum(statements.total_exec_time + statements.total_plan_time) over()) * 100, 'FM90D0') || '%'  as prop_total_time
```

**Review by alaister — APPROVED** [review]
LGTM!

**Comment by MildTomato** [comment]
Merging this in @kiwicopple @KevinBrolly  - assumed that was OK since they were non critical doc changes.
Can edit them more while in prod.

**Pull request merged**